### PR TITLE
Implements #3691.

### PR DESCRIPTION
**Pull request by dom96**

### CODE REVIEW

File: compiler/ast.nim [hunk 1]
```diff
@@ -1584,6 +1584,22 @@ proc skipStmtList*(n: PNode): PNode =
...
+    result = newType(tyRef, typ.owner)
+    rawAddSon(result, typ)
+
+proc toObject*(typ: PType): PType =
+  ## If ``typ`` is a tyRef then its immediate son is returned (which in many
+  ## cases should be a ``tyObject``).
+  ## Otherwise ``typ`` is simply returned as-is.
+  result = typ
+  if result.kind == tyRef:
+    result = result.sons[0]
```
Comment (Araq) [inline compiler/ast.nim]: that needs to ``lastSon``

File: compiler/semstmts.nim [hunk 2]
```diff
@@ -263,36 +263,63 @@ proc semTry(c: PContext, n: PNode): PNode =
...
         length = a.sonsLen
 
+      # Iterate through each exception type in the except branch.
       for j in countup(0, length-2):
-        var typ = semTypeNode(c, a.sons[j], nil)
-        if typ.kind == tyRef: typ = typ.sons[0]
+        var typeNode = a.sons[j] # e.g. `Exception`
+        var symbolNode: PNode = nil # e.g. `foobar`
+        # Handle the case where the `Exception as foobar` syntax is used.
+        if typeNode.kind == nkInfix:
```
Comment (Araq) [inline compiler/semstmts.nim]: that misses a check like ``and considerQuotedIdent(typeNode[0]).s == "as"``

File: compiler/transf.nim [hunk 3]
```diff
@@ -701,6 +701,36 @@ proc transformCall(c: PTransf, n: PNode): PTransNode =
     else:
       result = s.PTransNode
 
+proc transformExceptBranch(c: PTransf, n: PNode): PTransNode =
+  result = transformSons(c, n)
+  if n[0].kind == nkInfix:
```
Comment (Araq) [inline compiler/transf.nim]: again, missing check for "as"

File: compiler/semgnrc.nim [hunk 4]
```diff
@@ -339,8 +339,17 @@ proc semGenericStmt(c: PContext, n: PNode,
       checkMinSonsLen(a, 1)
       var L = sonsLen(a)
       for j in countup(0, L-2):
-        a.sons[j] = semGenericStmt(c, a.sons[j], flags+{withinTypeDesc}, ctx)
+        debug(a.sons[j])
```
Comment (Araq) [inline compiler/semgnrc.nim]: 'debug' code left

File: compiler/lookups.nim [hunk 5]
```diff
@@ -434,3 +434,6 @@ proc pickSym*(c: PContext, n: PNode; kind: TSymKind;
     if a.kind == kind and flags <= a.flags:
       return a
     a = nextOverloadIter(o, c, n)
+
+proc isExceptAs*(n: PNode): bool =
```
Comment (Araq) [inline compiler/lookups.nim]: well that tests for ``isInfixAs``

File: compiler/semgnrc.nim (hunk 4 above)
Comment (dom96) [inline compiler/semgnrc.nim]: yes, note the ``WIP`` in my commit message.

**Pull request merged**